pymc/tests/distributions/test_multivariate.py

```diff
@@ -1367,6 +1391,100 @@ def test_issue_3706(self):
+
+        # to test forward graph
+        random_samples = pm.draw(
+            v,
+            draws=10,
+        )
+
+        assert s.posterior.v.shape == (1, 10, len(COORDS["regions"]), len(COORDS["answers"]))
+
+        if not isinstance(zerosum_axes, (list, tuple)):
+            zerosum_axes = [zerosum_axes]
+
+        if isinstance(zerosum_axes[0], str):
+            for ax in zerosum_axes:
+                for samples in [
+                    s.posterior.v.mean(dim=ax),
+                    random_samples.mean(axis=dims.index(ax) + 1),
+                ]:
+                    assert np.isclose(
+                        samples, 0
+                    ).all(), f"{ax} is a zerosum_axis but is not summing to 0 across all samples."
+
+            nonzero_axes = list(set(dims).difference(zerosum_axes))
+            if nonzero_axes:
+                for ax in nonzero_axes:
+                    for samples in [
+                        s.posterior.v.mean(dim=ax),
+                        random_samples.mean(axis=dims.index(ax) + 1),
+                    ]:
+                        assert not np.isclose(
+                            samples, 0
+                        ).all(), f"{ax} is not a zerosum_axis, but is nonetheless summing to 0 across all samples."
+
+        else:
+            for ax in zerosum_axes:
+                if ax < 0:
+                    assert np.isclose(
+                        s.posterior.v.mean(axis=ax), 0
+                    ).all(), f"{ax} is a zerosum_axis but is not summing to 0 across all samples."
+                else:
+                    ax = ax + 2  # because 'chain' and 'draw' are added as new axes after sampling
+                    assert np.isclose(
+                        s.posterior.v.mean(axis=ax), 0
+                    ).all(), f"{ax} is a zerosum_axis but is not summing to 0 across all samples."
+
+    @pytest.mark.parametrize(
+        "dims, zerosum_axes",
+        [
+            (("regions", "answers"), 2),
+            (("regions", "answers"), (0, -2)),
+        ],
+    )
+    def test_zsn_fail_axis(self, dims, zerosum_axes):
+        if isinstance(zerosum_axes, (list, tuple)):
+            with pytest.raises(ValueError, match="repeated axis"):
+                with pm.Model(coords=COORDS) as m:
+                    _ = pm.ZeroSumNormal("v", dims=dims, zerosum_axes=zerosum_axes)
+        else:
+            with pytest.raises(AxisError, match="out of bounds"):
+                with pm.Model(coords=COORDS) as m:
+                    _ = pm.ZeroSumNormal("v", dims=dims, zerosum_axes=zerosum_axes)
+
+    def test_zsn_change_dist_size(self):
+        base_dist = pm.ZeroSumNormal.dist(shape=(4, 9))
+
+        new_dist = change_dist_size(base_dist, new_size=(5, 3), expand=False)
+        assert new_dist.eval().shape == (5, 3)
+
+        new_dist = change_dist_size(base_dist, new_size=(5, 3), expand=True)
+        assert new_dist.eval().shape == (5, 3, 4, 9)
+
+
```

Comment:
We also need to add a test from the logp against the baseline singular MvNormal

Reply:
Meaning that the ZSN logp should be equal to the MvNormal's?

Reply:
It’s equal to a special form of MvNormal. The thing is that the covariance matrix of the MvNormal is singular and you can’t compute its logp using our pymc dist. I’ll pull up the original comment from the old PR where I talked about this

Reply:
Here’s the link to the old [thread](https://github.com/pymc-devs/pymc/pull/4776#discussion_r800245250) about the logp

Reply:
Thanks @lucianopaz ! So this means we should test our ZSN logp against the one you posted in the old thread (`logp(value, sigma)`)?
If yes, then that means we can only test the case where only one axis sums to zero?

Reply:
I’ll send you a notebook I have where this is tested against multiple axes

Reply:
Oh cool, thanks @lucianopaz !